DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/EncodedBlocks.h

```diff
@@ -808,59 +815,79 @@ void EncodedBlocks<H, N, W>::encode(const S_IT srcBegin,     // iterator begin o
-    bl->setNData(dataSize);
-    bl->realignBlock();
+    storageBuffer_t* const blockBufferBegin = thisBlock->getCreateData();
+    const size_t maxBufferSize = thisBlock->registry->getFreeSize(); // note: "this" might be not valid after expandStorage call!!!
+    const auto encodedMessageEnd = encoder->process(srcBegin, srcEnd, blockBufferBegin, literals);
+    rans::utils::checkBounds(encodedMessageEnd, blockBufferBegin + maxBufferSize);
+    dataSize = encodedMessageEnd - thisBlock->getData();
+    thisBlock->setNData(dataSize);
+    thisBlock->realignBlock();
     // update the size claimed by encode message directly inside the block
-    // store incompressible symbols if any
 
-    int literalSize = 0;
-    if (literals.size()) {
-      literalSize = (literals.size() * sizeof(STYP)) / sizeof(stream_t) + (sizeof(STYP) < sizeof(stream_t));
-      expandStorage(literalSize);
-      bl->storeLiterals(literalSize, reinterpret_cast<const stream_t*>(literals.data()));
-    }
-    *meta = Metadata{messageLength, literals.size(), sizeof(uint64_t), sizeof(stream_t), static_cast<uint8_t>(encoder->getSymbolTablePrecision()), opt,
-                     encoder->getMinSymbol(), encoder->getMaxSymbol(), dictSize, dataSize, literalSize};
+    // store incompressible symbols if any
+    const size_t nLiteralSymbols = [&]() {
+      const size_t nSymbols = literals.size();
+      if (!literals.empty()) {
+        // introduce padding in case literals don't align;
+        const size_t nPaddingElements = (nSymbols * sizeof(input_t)) % sizeof(storageBuffer_t);
+        literals.insert(literals.end(), nPaddingElements, {});
+
+        const size_t literalStorageSize = (literals.size() * sizeof(input_t)) / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
+        expandStorage(literalStorageSize);
+        thisBlock->storeLiterals(literalStorageSize, reinterpret_cast<const storageBuffer_t*>(literals.data()));
+      }
+      return nSymbols;
+    }();
+
+    *thisMetadata = Metadata{messageLength,
+                             literals.size(),
+                             sizeof(ransState_t),
+                             sizeof(ransStream_t),
+                             static_cast<uint8_t>(encoder->getSymbolTablePrecision()),
+                             opt,
+                             encoder->getMinSymbol(),
+                             encoder->getMaxSymbol(),
+                             static_cast<int32_t>(frequencyTable.size()),
+                             dataSize,
+                             static_cast<int32_t>(nLiteralSymbols)};
 
   } else { // store original data w/o EEncoding
-    const size_t szb = messageLength * sizeof(STYP);
-    const int dataSize = szb / sizeof(stream_t) + (sizeof(STYP) < sizeof(stream_t));
+    const size_t messageSizeBytes = messageLength * sizeof(input_t);
+    const int dataSize = messageSizeBytes / sizeof(storageBuffer_t) + (sizeof(input_t) < sizeof(storageBuffer_t));
```

Comment:
same comment here: you add unneeded slot even if when there is no alignment problem.